hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/util/PendingAirdropUpdater.java

```diff
@@ -0,0 +1,125 @@
+ * and {@code nextAirdrop()}, also sender's account updates of {@code headPendingAirdropId()} and
+ * {@code numberPendingAirdrops()}
+ */
+public class PendingAirdropUpdater {
+    @NonNull
+    private final WritableAirdropStore pendingAirdropStore;
+
+    @NonNull
+    private final WritableAccountStore accountStore;
+
+    public PendingAirdropUpdater(
+            @NonNull WritableAirdropStore pendingAirdropStore, @NonNull WritableAccountStore accountStore) {
+        this.pendingAirdropStore = pendingAirdropStore;
+        this.accountStore = accountStore;
+    }
+
+    public void removePendingAirdrops(@NonNull final List<PendingAirdropId> airdropsToRemove) {
+        final Map<AccountID, Account> updatedSenders = new HashMap<>();
+        final Map<PendingAirdropId, AccountPendingAirdrop> updatedAirdrops = new HashMap<>();
+        // calculate state changes
+        for (PendingAirdropId id : airdropsToRemove) {
+            removePendingAirdropAndUpdateStores(id, updatedSenders, updatedAirdrops);
+        }
+
+        // commit updates
+        updatedSenders.forEach((accountID, account) -> accountStore.put(account));
+        updatedAirdrops.forEach(pendingAirdropStore::put);
+        airdropsToRemove.forEach(pendingAirdropStore::remove);
+    }
+
+    private void removePendingAirdropAndUpdateStores(
+            final PendingAirdropId airdropId,
+            final Map<AccountID, Account> updatedSenders,
+            final Map<PendingAirdropId, AccountPendingAirdrop> updatedAirdrops) {
+
+        final var senderId = airdropId.senderIdOrThrow();
+        var senderAccount = updatedSenders.containsKey(senderId)
+                ? updatedSenders.get(senderId)
+                : requireNonNull(accountStore.getAccountById(senderId));
+
+        final var airdrop = updatedAirdrops.containsKey(airdropId)
+                ? updatedAirdrops.get(airdropId)
+                : pendingAirdropStore.getForModify(airdropId);
+        validateTrue(airdrop != null, INVALID_TRANSACTION_BODY);
+
+        // updated sender's head of pending airdrops
+        if (airdropId.equals(senderAccount.headPendingAirdropId())) {
+            final var updatedSender = senderAccount
+                    .copyBuilder()
+                    .headPendingAirdropId(airdrop.nextAirdrop())
+                    .build();
+            updatedSenders.put(updatedSender.accountId(), updatedSender);
+        }
+
+        // update pending airdrops links
+        final var prevAirdropId = airdrop.previousAirdrop();
+        final var nextAirdropId = airdrop.nextAirdrop();
+        if (prevAirdropId != null) {
+            final var prevAccountAirdrop = updatedAirdrops.containsKey(prevAirdropId)
+                    ? updatedAirdrops.get(prevAirdropId)
+                    : pendingAirdropStore.getForModify(prevAirdropId);
+            validateTrue(prevAccountAirdrop != null, INVALID_TRANSACTION_BODY);
+            final var prevAirdropToUpdate =
```

Comment:
Link updates should probably be in one place and re-used between airdrops, claim and cancel

Reply:
Agree. 
Thats why I created this class. We will add it in the cancel PR too.